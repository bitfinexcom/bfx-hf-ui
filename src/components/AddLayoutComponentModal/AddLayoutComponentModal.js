import React from 'react'
import _isEmpty from 'lodash/isEmpty'

import Modal from '../../ui/Modal'
import Button from '../../ui/Button'
import Dropdown from '../../ui/Dropdown'
import {
  COMPONENT_TYPES, COMPONENT_LABELS,
} from '../GridLayout/GridLayout.helpers'

import { propTypes, defaultProps } from './AddLayoutComponentModal.props'
import './style.css'

export default class AddLayoutComponentModal extends React.Component {
  static propTypes = propTypes
  static defaultProps = defaultProps

  state = {
    componentType: COMPONENT_LABELS.CHART,
    error: '',
  }

  constructor(props) {
    super(props)

    this.onComponentTypeChange = this.onComponentTypeChange.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
  }

  onComponentTypeChange(componentType) {
    this.setState(() => ({ componentType }))
  }

  onSubmit() {
    const { componentType } = this.state
    const { onSubmit, onClose } = this.props

    if (_isEmpty(componentType) || !COMPONENT_LABELS[componentType]) {
      this.setState(() => ({ error: 'Invalid Component' }))
      return
    }

    onSubmit(componentType)
    onClose()
  }

  render() {
    const { onClose } = this.props
    const { componentType, error } = this.state

    return (
      <Modal
        onClose={onClose}
        className='hfui-addlayoutcomponentmodal__wrapper'
        label='Add Component'
        actions={(
          <Button
            green
            onClick={this.onSubmit}
            label={[
              <i key='icon' className='icon-plus' />,
              <p key='text'>Add Component</p>,
            ]}
          />
        )}
      >
        <Dropdown
          value={componentType}
          onChange={this.onComponentTypeChange}
          options={Object.values(COMPONENT_TYPES).map(type => ({
            label: COMPONENT_LABELS[type],
            value: type,
          }))}
        />

        {!_isEmpty(error) && (
          <p className='error'>{error}</p>
        )}
      </Modal>
    )
  }
}
