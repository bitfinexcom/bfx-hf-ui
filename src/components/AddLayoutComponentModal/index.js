import AddLayoutComponentModal from './AddLayoutComponentModal'

export default AddLayoutComponentModal
