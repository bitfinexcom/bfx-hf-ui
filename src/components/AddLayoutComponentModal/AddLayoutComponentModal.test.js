import React from 'react'
import { configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import AddLayoutComponentModal from './AddLayoutComponentModal'
import {
  COMPONENT_TYPES, COMPONENT_LABELS
} from '../GridLayout/GridLayout.helpers'

const LABELS = Object.values(COMPONENT_LABELS)

configure({ adapter: new Adapter() })

describe('AddLayoutComponentModal', () => {
  test('it renders a dropdown with one of each component type', () => {
    const component = mount((
      <AddLayoutComponentModal
        onSubmit={() => {}}
        onClose={() => {}}
      />
    ))

    expect(component).toMatchSnapshot()
    component.find('.hfui-dropdown__button').simulate('click')

    const dropdown = component.find('.hfui-dropdown__dropdown')

    LABELS.forEach((label) => {
      expect(dropdown.contains(label)).toBe(true)
    })
  })

  test('it shows an error if attempting to submit an unknown component type', () => {
    const component = mount((
      <AddLayoutComponentModal
        onSubmit={() => {}}
        onClose={() => {}}
      />
    ))

    expect(component.find('.error').length).toBe(0)
    expect(component.state('error')).toBe('')

    component.instance().onSubmit('nonexistent-type')

    expect(component.state('error')).not.toBe('')
  })

  test('it calls onSubmit() with the selected component type when confirmed', (done) => {
    const component = mount((
      <AddLayoutComponentModal
        onClose={() => {}}
        onSubmit={(componentType) => {
          expect(componentType).toBe(COMPONENT_TYPES.ORDER_BOOK)
          done()
        }}
      />
    ))

    component.find('.hfui-dropdown__button').simulate('click')

    const obOption = component.find('.hfui-dropdown__dropdown li').findWhere(c => (
      c.text() === COMPONENT_LABELS.ORDER_BOOK && c.prop('onClick')
    ))

    obOption.simulate('click')

    component.find('.hfui-button.green').simulate('click')
  })
})
