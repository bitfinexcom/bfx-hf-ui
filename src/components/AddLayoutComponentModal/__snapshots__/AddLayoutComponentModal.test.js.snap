// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddLayoutComponentModal it renders a dropdown with one of each component type 1`] = `
<AddLayoutComponentModal
  onClose={[Function]}
  onSubmit={[Function]}
>
  <Modal
    actions={
      <Button
        className=""
        green={true}
        label={
          Array [
            <i
              className="icon-plus"
            />,
            <p>
              Add Component
            </p>,
          ]
        }
        onClick={[Function]}
      />
    }
    className="hfui-addlayoutcomponentmodal__wrapper"
    label="Add Component"
    onClose={[Function]}
  >
    <div
      className="hfui-modal__wrapper"
      onClick={[Function]}
    >
      <div
        className="hfui-modal__content hfui-addlayoutcomponentmodal__wrapper"
        onClick={[Function]}
      >
        <div
          className="hfui-modal__header"
        >
          <p
            className="hfui-modal__label"
          >
            Add Component
          </p>
          <i
            className="icon-cancel"
            onClick={[Function]}
          />
        </div>
        <OnClickOutside(Dropdown)
          eventTypes={
            Array [
              "mousedown",
              "touchstart",
            ]
          }
          excludeScrollbar={false}
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "Chart",
                "value": "CHART",
              },
              Object {
                "label": "Order Book",
                "value": "ORDER_BOOK",
              },
              Object {
                "label": "Order Form",
                "value": "ORDER_FORM",
              },
              Object {
                "label": "Trades Table",
                "value": "TRADES_TABLE",
              },
              Object {
                "label": "Positions Table",
                "value": "POSITIONS_TABLE",
              },
              Object {
                "label": "Balances Table",
                "value": "BALANCES_TABLE",
              },
              Object {
                "label": "Algo Orders Table",
                "value": "ALGO_ORDERS_TABLE",
              },
              Object {
                "label": "Atomic Orders Table",
                "value": "ATOMIC_ORDERS_TABLE",
              },
              Object {
                "label": "Order History Table",
                "value": "ORDER_HISTORY_TABLE",
              },
              Object {
                "label": "Trading State Panel",
                "value": "TRADING_STATE_PANEL",
              },
            ]
          }
          outsideClickIgnoreClass="ignore-react-onclickoutside"
          preventDefault={false}
          stopPropagation={false}
          value="Chart"
        >
          <Dropdown
            disableOnClickOutside={[Function]}
            enableOnClickOutside={[Function]}
            eventTypes={
              Array [
                "mousedown",
                "touchstart",
              ]
            }
            onChange={[Function]}
            options={
              Array [
                Object {
                  "label": "Chart",
                  "value": "CHART",
                },
                Object {
                  "label": "Order Book",
                  "value": "ORDER_BOOK",
                },
                Object {
                  "label": "Order Form",
                  "value": "ORDER_FORM",
                },
                Object {
                  "label": "Trades Table",
                  "value": "TRADES_TABLE",
                },
                Object {
                  "label": "Positions Table",
                  "value": "POSITIONS_TABLE",
                },
                Object {
                  "label": "Balances Table",
                  "value": "BALANCES_TABLE",
                },
                Object {
                  "label": "Algo Orders Table",
                  "value": "ALGO_ORDERS_TABLE",
                },
                Object {
                  "label": "Atomic Orders Table",
                  "value": "ATOMIC_ORDERS_TABLE",
                },
                Object {
                  "label": "Order History Table",
                  "value": "ORDER_HISTORY_TABLE",
                },
                Object {
                  "label": "Trading State Panel",
                  "value": "TRADING_STATE_PANEL",
                },
              ]
            }
            outsideClickIgnoreClass="ignore-react-onclickoutside"
            preventDefault={false}
            stopPropagation={false}
            value="Chart"
          >
            <div
              className="hfui-dropdown__wrapper"
            >
              <div
                className="hfui-dropdown__button-wrapper"
              >
                <div
                  className="hfui-dropdown__button"
                  onClick={[Function]}
                >
                  <p
                    className=""
                  >
                    Select an option
                  </p>
                  <i
                    className="icon-arrow-down-passive"
                  />
                </div>
              </div>
            </div>
          </Dropdown>
        </OnClickOutside(Dropdown)>
        <div
          className="hfui-modal__actions"
        >
          <Button
            className=""
            green={true}
            label={
              Array [
                <i
                  className="icon-plus"
                />,
                <p>
                  Add Component
                </p>,
              ]
            }
            onClick={[Function]}
          >
            <button
              className="hfui-button green"
              onClick={[Function]}
              type="button"
            >
              <i
                className="icon-plus"
                key="icon"
              />
              <p
                key="text"
              >
                Add Component
              </p>
            </button>
          </Button>
        </div>
      </div>
    </div>
  </Modal>
</AddLayoutComponentModal>
`;
